### PR TITLE
Add Array and typed arrays codecs

### LINKED ISSUE #87: Array encoders and decoders
There are no default `encoder()` and `decoder()` functions for array types (`Array<T>`, `DoubleArray`, etc.). It looks like these were not added because the array types do not have companion objects. I think helios should still have functions for them, though. For example, here is a `DoubleArray` decoder I threw together:

```kotlin
fun doubleArrayDecoder() = object : Decoder<DoubleArray> {
    override fun decode(value: Json): Either<DecodingError, DoubleArray> = binding {
        value.asJsArray()
            .toEither { ArrayDecodingError(value) }
            .bind()
            .value
            .map {
                it.asJsNumber()
                    .toEither { NumberDecodingError(it) }
                    .bind()
                    .decode(Double.decoder())
                    .bind()
            }
            .toDoubleArray()
    }
}
```

### PR DESCRIPTION
**Pull request by AdrianRaFo**
Fixes #87 
Add Codecs for the generic array and the typed specific version

Fix the Map decoder

Add tests for all the collections instances

### CODE REVIEW

**Comment by 47degdev** [comment]
***Hood benchmark comparison:***
:warning: `Decoding` (Threshold: 71.38626659803376)

Benchmark | Value
|--|--|
master_benchmark | 11228.296660731012
helios_benchmark | 11187.914381452074


:heavy_check_mark: `Decodingfromraw` (Threshold: 250.0)

Benchmark | Value
|--|--|
master_benchmark | 7052.770736463254
helios_benchmark | 7119.496005360624


:heavy_check_mark: `Parsing` (Threshold: 250.0)

Benchmark | Value
|--|--|
master_benchmark | 21138.582930960165
helios_benchmark | 22252.790038530064

File: helios-core/src/main/kotlin/helios/instances/instancesCollections.kt [hunk 1]
```diff
@@ -52,14 +51,233 @@ interface ListDecoderInstance<out A> : Decoder<List<A>> {
 
 }
 
+@extension
+interface ArrayEncoderInstance<A> : Encoder<Array<A>> {
+
+  fun encoderA(): Encoder<A>
+
+  override fun Array<A>.encode(): Json =
+    JsArray(map { encoderA().run { it.encode() } })
```
Comment (nomisRev) [inline helios-core/src/main/kotlin/helios/instances/instancesCollections.kt]:
```suggestion
    with(encoderA()) { JsArray(map { it.encode() }) }
```

We don't know for sure if `encoderA()` is memoized, so to be sure we should enable the syntax from the most outer scope.

File: helios-core/src/main/kotlin/helios/instances/instancesCollections.kt [hunk 2]
```diff
@@ -52,14 +51,233 @@ interface ListDecoderInstance<out A> : Decoder<List<A>> {
...
+
+}
+
+@extension
+interface FloatArrayDecoderInstance : Decoder<FloatArray> {
+
+  override fun decode(value: Json): Either<DecodingError, FloatArray> =
+    value.asJsArray().toList()
+      .flatMap { arr ->
+        arr.value.map { Float.decoder().decode(it) }
```
Comment (nomisRev) [inline helios-core/src/main/kotlin/helios/instances/instancesCollections.kt]: FYI: these should be memoized by `@extension` so non-issue.

File: helios-core/src/main/kotlin/helios/instances/instancesCollections.kt [hunk 3]
```diff
@@ -52,14 +51,233 @@ interface ListDecoderInstance<out A> : Decoder<List<A>> {
...
 @extension
 interface MapEncoderInstance<A, B> : Encoder<Map<A, B>> {
 
   fun keyEncoderA(): KeyEncoder<A>
   fun encoderB(): Encoder<B>
 
   override fun Map<A, B>.encode(): Json =
-    JsObject(this.map { (key, value) -> (keyEncoderA().run { key.keyEncode() } to encoderB().run { value.encode() }) }.toMap())
+    JsObject(this.map { (key, value) ->
+      (keyEncoderA().run { key.keyEncode().value.toString() } to encoderB().run { value.encode() })
```
Comment (nomisRev) [inline helios-core/src/main/kotlin/helios/instances/instancesCollections.kt]: This syntax should be enabled from the most outer-scope.
```kotlin
with(keyEncoder()) {
  with(valueEncoder()) {
    JsObject(this.map { (key, value) -> ... })
 }
}
```

File: helios-core/src/main/kotlin/helios/instances/instancesCollections.kt [hunk 4]
```diff
@@ -80,8 +298,7 @@ interface MapDecoderInstance<A, B> : Decoder<Map<A, B>> {
   override fun decode(value: Json): Either<DecodingError, Map<A, B>> =
     value.asJsObject().fold({ ObjectDecodingError(value).left() }, { obj ->
       obj.value.map { (key, value) ->
-        val maybeKey: Either<DecodingError, A> =
-          Json.parseFromString(key).mapLeft { StringDecodingError(value) }.flatMap { keyDecoderA().keyDecode(it) }
+        val maybeKey: Either<DecodingError, A> = keyDecoderA().keyDecode(JsString(key))
```
Comment (nomisRev) [inline helios-core/src/main/kotlin/helios/instances/instancesCollections.kt]: Same.

**Comment by AdrianRaFo** [comment]
@nomisRev @fedefernandez ready for a second review

**Pull request merged**